modules/generator/registry/registry.go

```diff
@@ -208,20 +240,48 @@ func (r *ManagedRegistry) onAddMetricSeries(count uint32) bool {
 	return true
 }
 
-func (r *ManagedRegistry) onRemoveMetricSeries(count uint32) {
-	r.activeSeries.Sub(count)
+func (r *ManagedRegistry) onUpdateEntity(hash uint64) {
+	r.updatedSinceLastCollectionMtx.Lock()
+	r.updatedSinceLastCollection[hash] = struct{}{}
+	r.updatedSinceLastCollectionMtx.Unlock()
+}
 
+func (r *ManagedRegistry) onRemoveEntity(count uint32) {
+	r.activeSeries.Sub(count)
 	r.metricTotalSeriesRemoved.Add(float64(count))
 }
 
 func (r *ManagedRegistry) CollectMetrics(ctx context.Context) {
+	r.updatedSinceLastCollectionMtx.RLock()
+	hashes := maps.Clone(r.updatedSinceLastCollection)
+	r.updatedSinceLastCollection = make(map[uint64]struct{})
+	r.updatedSinceLastCollectionMtx.RUnlock()
+
 	r.metricsMtx.RLock()
 	defer r.metricsMtx.RUnlock()
 
+	rejected, err := r.entityLimiter.TrackEntities(ctx, r.tenant, slices.Collect(maps.Keys(hashes)))
+	if err != nil {
+		r.limitLogger.Log("msg", "tracking entities failed", "err", err)
+	}
+
+	if len(rejected) > 0 {
+		r.limitLogger.Log("msg", "max active entities reached", "rejected", len(rejected), "updated_since_last_collection", len(hashes))
+	}
+
+	rejectedMap := make(map[uint64]struct{})
+	for _, hash := range rejected {
+		rejectedMap[hash] = struct{}{}
+	}
+
 	var activeSeries int
 	var seriesDemand int
 
 	for _, m := range r.metrics {
+		m.deleteFunc(func(hash uint64, _ int64) bool {
```

Comment:
ok, I think I'm seeing. this is another way to cleanup metrics. we still have the time based cleanup below, but this is how you are enforcing the "max active entities" concept. on each collection loop you are removing all series whose "entity" is not in the kept group.

Reply:
That's right, it could also be another function, but this felt more right to me. It's the same interface as `slices.DeleteFunc` and is the pattern we use in adaptive metrics' series tracking to manage garbage collection.

Reply:
One alternative would be to keep `removeStaleSeries` and instead add another `delete(hash)` method. WDYT?